File: habitat3-exploration/ros_workspace/src/explorer_bridge/explorer_bridge/known_pose_pc_mapper_node.py
#!/usr/bin/env python3
"""Build /grid_map from /depth_data + stamped /odom (privileged pose; no SLAM).

Point-cloud-style mapping: ray-carve free space from the full depth FOV, but
mark occupied only for hits between floor and robot height (~1 m).
"""

from __future__ import annotations

import math
from collections import OrderedDict, deque

import rclpy
from nav_msgs.msg import MapMetaData, OccupancyGrid, Odometry
from rclpy.node import Node
from rclpy.qos import (
    DurabilityPolicy,
    HistoryPolicy,
    QoSProfile,
    ReliabilityPolicy,
    qos_profile_sensor_data,
)
from sensor_msgs.msg import CameraInfo, Image
from std_msgs.msg import Header

from explorer_bridge.image_utils import image_to_depth_array
from explorer_bridge.pc_to_occupancy import (
    DEFAULT_WALL_HEIGHT_MAX_M,
    DEFAULT_WALL_HEIGHT_MIN_M,
    depth_content_signature,
    integrate_depth_frame,
)
from explorer_bridge.scan_to_occupancy import (
    OccupancyMap,
    inflate_occupied,
    inflation_radius_cells,
    should_integrate_scan,
    should_integrate_with_tf,
    stamp_msg_to_ns,
)


def _yaw_from_odom(msg: Odometry) -> float:
    q = msg.pose.pose.orientation
    return math.atan2(2.0 * (q.w * q.z + q.x * q.y), 1.0 - 2.0 * (q.y * q.y + q.z * q.z))


class KnownPosePcMapperNode(Node):
    def __init__(self) -> None:
        super().__init__("known_pose_pc_mapper")
        self.declare_parameter("depth_topic", "/depth_data")
        self.declare_parameter("camera_info_topic", "/depth/camera_info")
        self.declare_parameter("odom_topic", "/odom")
        self.declare_parameter("grid_topic", "/grid_map")
        self.declare_parameter("map_frame", "map")
        self.declare_parameter("resolution", 0.05)
        self.declare_parameter("initial_size_m", 20.0)
        self.declare_parameter("publish_hz", 5.0)
        self.declare_parameter("odom_cache_size", 2048)
        self.declare_parameter("max_stamp_skew_sec", 0.0)
        self.declare_parameter("pending_depth_limit", 128)
        self.declare_parameter("obstacle_inflation_m", 0.10)
        self.declare_parameter("range_min", 0.1)
        self.declare_parameter("range_max", 10.0)
        self.declare_parameter("sensor_far", 50.0)
        self.declare_parameter("sat_eps", 0.5)
        self.declare_parameter("camera_z", 0.1)
        self.declare_parameter("wall_height_min_m", DEFAULT_WALL_HEIGHT_MIN_M)
        self.declare_parameter("wall_height_max_m", DEFAULT_WALL_HEIGHT_MAX_M)
        self.declare_parameter("subsample", 4)

        depth_topic = str(self.get_parameter("depth_topic").value)
        info_topic = str(self.get_parameter("camera_info_topic").value)
        odom_topic = str(self.get_parameter("odom_topic").value)
        grid_topic = str(self.get_parameter("grid_topic").value)
        self._map_frame = str(self.get_parameter("map_frame").value)
        resolution = float(self.get_parameter("resolution").value)
        initial_size = float(self.get_parameter("initial_size_m").value)
        publish_hz = max(0.2, float(self.get_parameter("publish_hz").value))
        self._cache_size = max(8, int(self.get_parameter("odom_cache_size").value))
        skew_sec = max(0.0, float(self.get_parameter("max_stamp_skew_sec").value))
        self._max_skew_ns = int(skew_sec * 1_000_000_000)
        pending_limit = max(1, int(self.get_parameter("pending_depth_limit").value))
        inflation_m = max(0.0, float(self.get_parameter("obstacle_inflation_m").value))
        self._inflate_cells = inflation_radius_cells(resolution, inflation_m)

        self._range_min = float(self.get_parameter("range_min").value)
        self._range_max = float(self.get_parameter("range_max").value)
        self._sensor_far = float(self.get_parameter("sensor_far").value)
        self._sat_eps = float(self.get_parameter("sat_eps").value)
        self._camera_z = float(self.get_parameter("camera_z").value)
        self._wall_height_min = float(self.get_parameter("wall_height_min_m").value)
        self._wall_height_max = float(self.get_parameter("wall_height_max_m").value)
        self._subsample = max(1, int(self.get_parameter("subsample").value))

        self._fx = 320.0
        self._fy = 320.0
        self._cx = 320.0
        self._cy = 240.0
        self._have_info = False

        self._grid = OccupancyMap(resolution=resolution, initial_size_m=initial_size)
        self._last_sig: tuple | None = None
        self._last_yaw: float | None = None
        self._odom_by_stamp: OrderedDict[int, tuple[float, float, float]] = OrderedDict()
        self._pending_depth: deque[Image] = deque(maxlen=pending_limit)

        map_qos = QoSProfile(
            depth=1,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.TRANSIENT_LOCAL,
            history=HistoryPolicy.KEEP_LAST,
        )
        self._pub = self.create_publisher(OccupancyGrid, grid_topic, map_qos)
        self.create_subscription(Odometry, odom_topic, self._odom_cb, 50)
        self.create_subscription(CameraInfo, info_topic, self._info_cb, 10)
        self.create_subscription(Image, depth_topic, self._depth_cb, qos_profile_sensor_data)
        self.create_timer(1.0 / publish_hz, self._publish)
        self.get_logger().info(
            f"Known-pose PC mapper: {depth_topic} + {odom_topic} → {grid_topic} "
            f"(wall band [{self._wall_height_min:.2f}, {self._wall_height_max:.2f}] m, "
            f"subsample={self._subsample})"
        )

    def _info_cb(self, msg: CameraInfo) -> None:
        self._fx = float(msg.k[0])
        self._fy = float(msg.k[4])
        self._cx = float(msg.k[2])
        self._cy = float(msg.k[5])
        self._have_info = True

    def _odom_cb(self, msg: Odometry) -> None:
        stamp_ns = stamp_msg_to_ns(msg.header.stamp)
        pose = (
            float(msg.pose.pose.position.x),
            float(msg.pose.pose.position.y),
            _yaw_from_odom(msg),
        )
        self._odom_by_stamp[stamp_ns] = pose
        while len(self._odom_by_stamp) > self._cache_size:
            self._odom_by_stamp.popitem(last=False)
        self._drain_pending_depth()

    def _depth_cb(self, msg: Image) -> None:
        if not self._have_info or msg.encoding != "32FC1":
            return
        if not self._try_integrate_depth(msg):
            self._pending_depth.append(msg)
            self.get_logger().warn(
                "No /odom with exact depth stamp "
                f"(cache={len(self._odom_by_stamp)}; deferred)",
                throttle_duration_sec=2.0,
            )

    def _drain_pending_depth(self) -> None:
        if not self._pending_depth:
            return
        remaining: list[Image] = []
        while self._pending_depth:
            img = self._pending_depth.popleft()
            stamp_ns = stamp_msg_to_ns(img.header.stamp)
            if stamp_ns in self._odom_by_stamp:
                self._try_integrate_depth(img)
                continue
            if self._odom_by_stamp:
                oldest = next(iter(self._odom_by_stamp))
                if stamp_ns < oldest:
                    continue
            remaining.append(img)
        self._pending_depth.clear()
        self._pending_depth.extend(remaining)

    def _try_integrate_depth(self, msg: Image) -> bool:
        stamp_ns = stamp_msg_to_ns(msg.header.stamp)
        pose = self._odom_by_stamp.get(stamp_ns)
        if self._max_skew_ns > 0 and pose is None:
            from explorer_bridge.scan_to_occupancy import find_pose_for_stamp

            cache = [(ts, *xyyaw) for ts, xyyaw in self._odom_by_stamp.items()]
            pose = find_pose_for_stamp(cache, stamp_ns, max_skew_ns=self._max_skew_ns)

        stamp_ok = pose is not None
        if not should_integrate_with_tf(stamp_lookup_ok=stamp_ok) or pose is None:
            return False

        depth = image_to_depth_array(msg)
        sig = depth_content_signature(depth, stride=self._subsample)
        robot_x, robot_y, yaw = pose
        if not should_integrate_scan(
            signature=sig,
            yaw=yaw,
            last_signature=self._last_sig,
            last_yaw=self._last_yaw,
        ):
            return True

        integrate_depth_frame(
            self._grid,
            depth,
            robot_x=robot_x,
            robot_y=robot_y,
            yaw=yaw,
            fx=self._fx,
            fy=self._fy,
            cx=self._cx,
            cy=self._cy,
            range_min=self._range_min,
            range_max=self._range_max,
            camera_z=self._camera_z,
            sensor_far=self._sensor_far,
            sat_eps=self._sat_eps,
            wall_height_min=self._wall_height_min,
            wall_height_max=self._wall_height_max,
            subsample=self._subsample,
        )
        self._last_sig = sig
        self._last_yaw = yaw
        return True

    def _publish(self) -> None:
        msg = OccupancyGrid()
        msg.header = Header(
            stamp=self.get_clock().now().to_msg(),
            frame_id=self._map_frame,
        )
        info = MapMetaData()
        info.resolution = self._grid.resolution
        info.width = self._grid.width
        info.height = self._grid.height
        info.origin.position.x = self._grid.origin_x
        info.origin.position.y = self._grid.origin_y
        info.origin.orientation.w = 1.0
        msg.info = info
        published = inflate_occupied(self._grid.data, radius_cells=self._inflate_cells)
        msg.data = published.flatten().tolist()
        self._pub.publish(msg)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = KnownPosePcMapperNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: habitat3-exploration/ros_workspace/src/explorer_bridge/src/known_pose_pc_mapper_node.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <deque>
#include <limits>
#include <map>
#include <memory>
#include <optional>
#include <string>
#include <vector>

#include <nav_msgs/msg/occupancy_grid.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <std_msgs/msg/header.hpp>

#include "explorer_bridge/occupancy_map.hpp"
#include "explorer_bridge/pc_occupancy.hpp"
#include "explorer_bridge/pose_gate.hpp"

namespace explorer_bridge
{
namespace
{

int64_t stampToNs(const builtin_interfaces::msg::Time & stamp)
{
  return static_cast<int64_t>(stamp.sec) * 1000000000LL + static_cast<int64_t>(stamp.nanosec);
}

double yawFromOdom(const nav_msgs::msg::Odometry & msg)
{
  const auto & q = msg.pose.pose.orientation;
  return std::atan2(
    2.0 * (q.w * q.z + q.x * q.y),
    1.0 - 2.0 * (q.y * q.y + q.z * q.z));
}

}  // namespace

class KnownPosePcMapperNode : public rclcpp::Node
{
public:
  KnownPosePcMapperNode()
  : Node("known_pose_pc_mapper")
  {
    declare_parameter("depth_topic", "/depth_data");
    declare_parameter("camera_info_topic", "/depth/camera_info");
    declare_parameter("odom_topic", "/odom");
    declare_parameter("grid_topic", "/grid_map");
    declare_parameter("map_frame", "map");
    declare_parameter("resolution", 0.05);
    declare_parameter("initial_size_m", 20.0);
    declare_parameter("publish_hz", 5.0);
    declare_parameter("odom_cache_size", 2048);
    declare_parameter("max_stamp_skew_sec", 0.0);
    declare_parameter("pending_depth_limit", 128);
    declare_parameter("obstacle_inflation_m", 0.05);
    declare_parameter("range_min", 0.1);
    declare_parameter("range_max", 10.0);
    declare_parameter("sensor_far", 50.0);
    declare_parameter("sat_eps", 0.5);
    declare_parameter("camera_z", 0.1);
    declare_parameter("wall_height_min_m", kDefaultWallHeightMin_m);
    declare_parameter("wall_height_max_m", kDefaultWallHeightMax_m);
    declare_parameter("subsample", 8);
    declare_parameter("min_pose_change_m", 0.25);
    declare_parameter("min_yaw_change_rad", 0.17);

    const auto depth_topic = get_parameter("depth_topic").as_string();
    const auto info_topic = get_parameter("camera_info_topic").as_string();
    const auto odom_topic = get_parameter("odom_topic").as_string();
    const auto grid_topic = get_parameter("grid_topic").as_string();
    map_frame_ = get_parameter("map_frame").as_string();
    const double resolution = get_parameter("resolution").as_double();
    const double initial_size = get_parameter("initial_size_m").as_double();
    const double publish_hz = std::max(0.2, get_parameter("publish_hz").as_double());
    cache_size_ = std::max(8, static_cast<int>(get_parameter("odom_cache_size").as_int()));
    const double skew_sec = std::max(0.0, get_parameter("max_stamp_skew_sec").as_double());
    max_skew_ns_ = static_cast<int64_t>(skew_sec * 1e9);
    pending_limit_ = std::max(1, static_cast<int>(get_parameter("pending_depth_limit").as_int()));
    const double inflation_m = std::max(0.0, get_parameter("obstacle_inflation_m").as_double());
    resolution_ = resolution;
    inflate_cells_ = inflationRadiusCells(resolution_, inflation_m);

    params_.range_min = get_parameter("range_min").as_double();
    params_.range_max = get_parameter("range_max").as_double();
    params_.sensor_far = get_parameter("sensor_far").as_double();
    params_.sat_eps = get_parameter("sat_eps").as_double();
    params_.camera_z = get_parameter("camera_z").as_double();
    params_.wall_height_min = get_parameter("wall_height_min_m").as_double();
    params_.wall_height_max = get_parameter("wall_height_max_m").as_double();
    params_.subsample = std::max(1, static_cast<int>(get_parameter("subsample").as_int()));
    min_pose_change_m_ = get_parameter("min_pose_change_m").as_double();
    min_yaw_change_rad_ = get_parameter("min_yaw_change_rad").as_double();

    grid_ = std::make_unique<OccupancyMap>(resolution, initial_size);

    rclcpp::QoS map_qos(1);
    map_qos.reliable();
    map_qos.transient_local();
    pub_ = create_publisher<nav_msgs::msg::OccupancyGrid>(grid_topic, map_qos);

    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      odom_topic, 50,
      [this](const nav_msgs::msg::Odometry::SharedPtr msg) {onOdom(msg);});
    info_sub_ = create_subscription<sensor_msgs::msg::CameraInfo>(
      info_topic, 10,
      [this](const sensor_msgs::msg::CameraInfo::SharedPtr msg) {onInfo(msg);});
    depth_sub_ = create_subscription<sensor_msgs::msg::Image>(
      depth_topic, rclcpp::SensorDataQoS(),
      [this](const sensor_msgs::msg::Image::SharedPtr msg) {onDepth(msg);});

    timer_ = create_wall_timer(
      std::chrono::duration<double>(1.0 / publish_hz),
      [this]() {publishGrid();});

    param_cb_ = add_on_set_parameters_callback(
      [this](const std::vector<rclcpp::Parameter> & params) {
        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;
        for (const auto & p : params) {
          if (p.get_name() == "obstacle_inflation_m") {
            const double m = std::max(0.0, p.as_double());
            inflate_cells_ = inflationRadiusCells(resolution_, m);
            RCLCPP_WARN(
              get_logger(),
              "obstacle_inflation_m set to %.3f (%d cells)", m, inflate_cells_);
          }
        }
        return result;
      });

    RCLCPP_INFO(
      get_logger(),
      "Known-pose PC mapper (C++): %s + %s -> %s (subsample=%d, pose_gate xy=%.2f yaw=%.2f)",
      depth_topic.c_str(), odom_topic.c_str(), grid_topic.c_str(),
      params_.subsample, min_pose_change_m_, min_yaw_change_rad_);
  }

private:
  void onInfo(const sensor_msgs::msg::CameraInfo::SharedPtr msg)
  {
    params_.K.fx = msg->k[0];
    params_.K.fy = msg->k[4];
    params_.K.cx = msg->k[2];
    params_.K.cy = msg->k[5];
    have_info_ = true;
  }

  void onOdom(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    const int64_t stamp_ns = stampToNs(msg->header.stamp);
    Pose2D pose{
      msg->pose.pose.position.x,
      msg->pose.pose.position.y,
      yawFromOdom(*msg)};
    odom_by_stamp_[stamp_ns] = pose;
    while (static_cast<int>(odom_by_stamp_.size()) > cache_size_) {
      odom_by_stamp_.erase(odom_by_stamp_.begin());
    }
    drainPending();
  }

  void onDepth(const sensor_msgs::msg::Image::SharedPtr msg)
  {
    if (!have_info_ || msg->encoding != "32FC1") {
      return;
    }
    if (!tryIntegrate(msg)) {
      pending_depth_.push_back(msg);
      while (static_cast<int>(pending_depth_.size()) > pending_limit_) {
        pending_depth_.pop_front();
      }
      RCLCPP_WARN_THROTTLE(
        get_logger(), *get_clock(), 2000,
        "No /odom with exact depth stamp (cache=%zu; deferred)",
        odom_by_stamp_.size());
    }
  }

  void drainPending()
  {
    if (pending_depth_.empty()) {
      return;
    }
    std::deque<sensor_msgs::msg::Image::SharedPtr> remaining;
    while (!pending_depth_.empty()) {
      auto img = pending_depth_.front();
      pending_depth_.pop_front();
      const int64_t stamp_ns = stampToNs(img->header.stamp);
      if (odom_by_stamp_.count(stamp_ns)) {
        tryIntegrate(img);
        continue;
      }
      if (!odom_by_stamp_.empty()) {
        const int64_t oldest = odom_by_stamp_.begin()->first;
        if (stamp_ns < oldest) {
          continue;
        }
      }
      remaining.push_back(img);
    }
    pending_depth_ = std::move(remaining);
  }

  std::optional<Pose2D> lookupPose(int64_t stamp_ns) const
  {
    const auto it = odom_by_stamp_.find(stamp_ns);
    if (it != odom_by_stamp_.end()) {
      return it->second;
    }
    if (max_skew_ns_ <= 0) {
      return std::nullopt;
    }
    std::optional<Pose2D> best;
    int64_t best_skew = std::numeric_limits<int64_t>::max();
    for (const auto & [ts, pose] : odom_by_stamp_) {
      const int64_t skew = std::llabs(ts - stamp_ns);
      if (skew < best_skew) {
        best_skew = skew;
        best = pose;
      }
    }
    if (!best || best_skew > max_skew_ns_) {
      return std::nullopt;
    }
    return best;
  }

  bool tryIntegrate(const sensor_msgs::msg::Image::SharedPtr & msg)
  {
    const int64_t stamp_ns = stampToNs(msg->header.stamp);
    auto pose = lookupPose(stamp_ns);
    if (!pose.has_value()) {
      return false;
    }

    if (!shouldIntegratePose(*pose, last_integrate_pose_, min_pose_change_m_, min_yaw_change_rad_)) {
      return true;
    }

    if (msg->step < msg->width * sizeof(float) ||
      msg->data.size() < static_cast<size_t>(msg->height) * msg->step)
    {
      RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 2000, "Malformed depth image");
      return true;
    }

    std::vector<float> depth(static_cast<size_t>(msg->height) * static_cast<size_t>(msg->width));
    for (uint32_t r = 0; r < msg->height; ++r) {
      const auto * row = reinterpret_cast<const float *>(
        msg->data.data() + static_cast<size_t>(r) * msg->step);
      std::memcpy(
        depth.data() + static_cast<size_t>(r) * msg->width,
        row,
        static_cast<size_t>(msg->width) * sizeof(float));
    }

    params_.robot_x = pose->x;
    params_.robot_y = pose->y;
    params_.yaw = pose->yaw;

    integrateDepthFrame(
      *grid_, depth.data(),
      static_cast<int>(msg->height), static_cast<int>(msg->width),
      params_);

    last_integrate_pose_ = *pose;
    return true;
  }

  void publishGrid()
  {
    auto published = inflateOccupied(*grid_, inflate_cells_);

    nav_msgs::msg::OccupancyGrid msg;
    msg.header.stamp = now();
    msg.header.frame_id = map_frame_;
    msg.info.resolution = static_cast<float>(grid_->resolution());
    msg.info.width = static_cast<uint32_t>(grid_->width());
    msg.info.height = static_cast<uint32_t>(grid_->height());
    msg.info.origin.position.x = grid_->originX();
    msg.info.origin.position.y = grid_->originY();
    msg.info.origin.orientation.w = 1.0;
    msg.data.assign(published.begin(), published.end());
    pub_->publish(msg);
  }

  std::string map_frame_;
  int cache_size_{2048};
  int64_t max_skew_ns_{0};
  int pending_limit_{128};
  int inflate_cells_{0};
  double resolution_{0.05};
  double min_pose_change_m_{0.25};
  double min_yaw_change_rad_{0.17};
  bool have_info_{false};
  IntegrateDepthParams params_{};
  std::unique_ptr<OccupancyMap> grid_;
  std::optional<Pose2D> last_integrate_pose_;
  std::map<int64_t, Pose2D> odom_by_stamp_;
  std::deque<sensor_msgs::msg::Image::SharedPtr> pending_depth_;

  rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr pub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr info_sub_;
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr depth_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr param_cb_;
};

}  // namespace explorer_bridge

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<explorer_bridge::KnownPosePcMapperNode>());
  rclcpp::shutdown();
  return 0;
}
